Render railway labels

littlebtc opened the pull request:
This will fix #505.

Rendering railway labels will make maps in some railway-centric area better, especially Japan.

I did some copy-and-paste work and make this pull request, hoping to get some feedback about how the label should be styled.

My two cents:

- Should the style of railway lines be changed to match the label?
- Subway under highways is mostly not rendered. Can we render that?

By following is some rendered images of the initial commit in Tokyo, Japan:

![2017-07-12 4 05 40](https://user-images.githubusercontent.com/82419/28088699-1e1410b4-66b9-11e7-92e2-da9731fc68f0.png)
<img width="748" alt="2017-07-12 4 00 16" src="https://user-images.githubusercontent.com/82419/28089420-8c421020-66bb-11e7-934c-0722fa33bc40.png">
![2017-07-12 4 04 20](https://user-images.githubusercontent.com/82419/28089478-b42993c4-66bb-11e7-90df-9fcd3ea191a1.png)

kocio-pl commented:
Could you provide before/after renderings? Tokyo is a dense city and I don't read Japanese, so it's hard for me to see where the changes are.

littlebtc commented:
After switching to Google Cloud , endless waiting for importing and rendering :joy:, my second approach is now ready to review.

Before/After images in London were made as requested. Before is the current rendering on OSM.org. Diff views were made with Kosmtik. I choose places near King's Cross Station to see the impacts on other labels.

The drawback of the change is obvious if you look at buildings and residential areas under the subway. 😢 

Z14
![london-z14](https://user-images.githubusercontent.com/82419/28282198-37f3e4c6-6b5c-11e7-8200-2c90e9be9b42.png)

Z15
![london-z15](https://user-images.githubusercontent.com/82419/28282289-892e194c-6b5c-11e7-85b2-25712c2646a8.png)

Z16
![london-z16](https://user-images.githubusercontent.com/82419/28282315-9718195e-6b5c-11e7-8626-08e97fe06983.png)

Z17
![london-z17](https://user-images.githubusercontent.com/82419/28282370-ca5466c4-6b5c-11e7-92a0-095cf4372096.png)

Z18
![london-z18](https://user-images.githubusercontent.com/82419/28282396-df8cad26-6b5c-11e7-94b1-ffe3d35ba204.png)

kocio-pl commented:
I'm not able to review SQL code, but I've just tested it and it looks good for me:

Warsaw, z18
Before
![epnazy5g](https://user-images.githubusercontent.com/5439713/28478991-e580e822-6e5a-11e7-9663-90743e246b43.png)
After
![pszuzuiu](https://user-images.githubusercontent.com/5439713/28479001-ea4d10e2-6e5a-11e7-81a9-a86e264c9980.png)

Was there any particular reason for you to start from z13?

littlebtc commented:
> Was there any particular reason for you to start from z13?
Since I copied and simplified styles from the highway name rendering, where `highway=motorway/trunk/primary` started render names at z13, and `highway=residential` started at z15.

Maybe started rendering from z11 can be a good idea, since road shields and motorway junction refs started at that.

But to make railways rendering more flexible, I want to find out a way to separated high-speed rail from normal one, since there is no simple way to determine this unless reading `maxspeed`.

littlebtc commented:
After visiting OpenRailwayMap, checking for current map rendering, I have pushed changes about rendering levels, label rendering in this approach is like below:

- `railway=rail` and `highspeed=yes` at z11
- `railway=rail` and `usage=main` at z11
- Most railways at z13, styling like `highway=primary`
- `railway=miniature`, service railways, disused, constructions at z15 like `highway=residential`

Before / After will be uploaded later.

kocio-pl commented:
Try to set export window in Kosmtik and just changing branches, so the resulting images will be directly comparable (I try opening in new tabs in my browser and fast switch between them).

littlebtc commented:
I will try to use my own Linux machine instead since my cloud machine has relatively small vCPU...

littlebtc commented:
Update: I have given up on checking `highspeed` and `usage` as it will took a hard work.

In the next images, all `railway=rail` will rendered on z11.

kocio-pl commented:
As I said in https://github.com/gravitystorm/openstreetmap-carto/issues/1959#issuecomment-317207988, database reload is not expected sooner than 5 months and as far as I understand changing `openstreetmap-carto.style` (or `openstreetmap-carto.lua`) would require database reload.

littlebtc commented:
Sorry for lots of waiting... Great London again.

**z11**
![z11-before](https://user-images.githubusercontent.com/82419/28495453-7a5d3fac-6f7f-11e7-8819-c53db8976704.jpg)
![z11-after](https://user-images.githubusercontent.com/82419/28495452-7a5aa4c2-6f7f-11e7-9299-c87b53c9514d.jpg)

**z12**
![z12-before](https://user-images.githubusercontent.com/82419/28495450-7a5449d8-6f7f-11e7-9cd3-d57cc938bafa.jpg)
![z12-after](https://user-images.githubusercontent.com/82419/28495449-7a528382-6f7f-11e7-83c0-b800655d036e.jpg)

**z13**
![z13-before](https://user-images.githubusercontent.com/82419/28495445-7a2c19e0-6f7f-11e7-840f-5175e833b835.jpg)
![z13-after](https://user-images.githubusercontent.com/82419/28495451-7a589ccc-6f7f-11e7-98f7-0daac6a8d99c.jpg)

**z14**
![z14-before](https://user-images.githubusercontent.com/82419/28495447-7a32328a-6f7f-11e7-9bf0-69704cba2ce6.jpg)
![z14-after](https://user-images.githubusercontent.com/82419/28495446-7a30aadc-6f7f-11e7-816e-321a8e541839.jpg)

**z15**
![z15-before](https://user-images.githubusercontent.com/82419/28495442-7a27216a-6f7f-11e7-9994-c7819043ac81.jpg)
![z15-after](https://user-images.githubusercontent.com/82419/28495448-7a4d4642-6f7f-11e7-90dd-e075bc41ddf7.jpg)

**z16**
![z16-before](https://user-images.githubusercontent.com/82419/28495443-7a2a6c12-6f7f-11e7-8d20-13ee58bcedaf.jpg)
![z16-after](https://user-images.githubusercontent.com/82419/28495444-7a2af592-6f7f-11e7-909d-0b31a225a2fb.jpg)

littlebtc commented:
Z11 in Tokyo, feeling like a different map.
That's what I originally wanted. :joy:

![2017-07-23 08-30-56](https://user-images.githubusercontent.com/82419/28495510-5aaa45b8-6f81-11e7-9e2c-9efd752621ad.png)
![2017-07-23 08-30-58](https://user-images.githubusercontent.com/82419/28495511-5aab8ebe-6f81-11e7-9157-cb5c920bf61a.png)

kocio-pl commented:
For me Tokyo shows that z11 is too early, map seems to be cluttered.

littlebtc commented:
It looks like I should generate some Tokyo before/after.

(Tokyo railway system are too far intensive)

littlebtc commented:
z12 to z17 before/after for Tokyo is here.
IMHO we can start render railways at z13 while removing the text halo from `railway=rail`.
Ideas?

For z12/z13, I centered at the central of the Tokyo.
**z12**
![z12-before](https://user-images.githubusercontent.com/82419/28495716-a1cb82a0-6f8a-11e7-8c17-2caec1b637d8.jpg)
![z12-after](https://user-images.githubusercontent.com/82419/28495717-a1cf734c-6f8a-11e7-8baf-040757639ad8.jpg)
**z13**
![z13-before](https://user-images.githubusercontent.com/82419/28495719-c3527e92-6f8a-11e7-82e1-f2625ff6bf2d.jpg)
![z13-after](https://user-images.githubusercontent.com/82419/28495720-c39a3e30-6f8a-11e7-8159-2fbe1f0b19cb.jpg)

For z14 and higher, I focused on Shinjuku, major suburbs and railway hub of Tokyo.
**z14**
![z14-before](https://user-images.githubusercontent.com/82419/28495723-e040040c-6f8a-11e7-961b-eb09aeed2077.jpg)
![z14-after](https://user-images.githubusercontent.com/82419/28495724-e04937c0-6f8a-11e7-8a00-9726e1e7e9ee.jpg)

**z15**
![z15-before](https://user-images.githubusercontent.com/82419/28495731-ed621b34-6f8a-11e7-888d-9fc2f9a4a80e.jpg)
![z15-after](https://user-images.githubusercontent.com/82419/28495730-ed61de30-6f8a-11e7-8e5c-28eab2615416.jpg)

**z16**
![z16-before](https://user-images.githubusercontent.com/82419/28495738-2787018a-6f8b-11e7-8f19-3e4f74f97313.jpg)
![z16-after](https://user-images.githubusercontent.com/82419/28495739-2789ea08-6f8b-11e7-8945-f13e7b44bd70.jpg)

**z17**
![z17-before](https://user-images.githubusercontent.com/82419/28495746-5fbc4d4e-6f8b-11e7-80b7-675e75847239.jpg)
![z17-after](https://user-images.githubusercontent.com/82419/28495747-5ffbea62-6f8b-11e7-94f2-1ccaca951a37.jpg)

kocio-pl commented on roads.mss:
```diff
@@ -3100,3 +3100,87 @@ tertiary is rendered from z10 and is not included in osm_planet_roads. */
+  [railway = 'rail'][zoom >= 11] {
+    text-name: "[name]";
+    text-fill: #333333;
+    text-size: 8;
```
We're trying to get rid of too small text sizes, see https://github.com/gravitystorm/openstreetmap-carto/issues/2209.

kocio-pl commented on roads.mss:
```diff
@@ -3100,3 +3100,87 @@ tertiary is rendered from z10 and is not included in osm_planet_roads. */
+      text-halo-fill: #cecece;
+    }
+    [zoom >= 14] {
+      text-size: 9;
```
The same text size problem as with https://github.com/gravitystorm/openstreetmap-carto/pull/2687/files#r129007651.

littlebtc commented:
Changes were made to reflect some comments and the review:

- SQL: Now using hstore to get `highspeed` and `usage`
- All font sizes started at 10
- New rendering levels:
  - `highspeed=yes` rail at z11
  - `usage=main` at z13
  - All others at z15 

Some `usage=main` font size still smaller in z15+, I will fix that.

littlebtc commented:
z11~z15 after this change:

![z11](https://user-images.githubusercontent.com/82419/28556060-44d43288-7136-11e7-8392-d3cefb2f7ec7.jpg)
![z12](https://user-images.githubusercontent.com/82419/28556062-44d61a94-7136-11e7-8e94-39c9138be99d.jpg)
![Uploading z13.jpg…]()
![z14](https://user-images.githubusercontent.com/82419/28556063-44d6d588-7136-11e7-96c8-4f9f4c654379.jpg)
![z15](https://user-images.githubusercontent.com/82419/28556064-45a24010-7136-11e7-823e-e36cf39968f9.jpg)

kocio-pl commented on roads.mss:
```diff
@@ -3100,3 +3100,81 @@ tertiary is rendered from z10 and is not included in osm_planet_roads. */
   }
 }
+#railways-text-name {
+  /* Others started from z15. */
```
I think subway, tram and other city transport labels should be rendered much later (like z18-z19), since such cities tend to be cluttered with labels anyway.

littlebtc replied:
I think z18 is a little bit late; anyway I will try to render New York to see how does it look.

littlebtc replied:
In Manhattan NYC, z14 ~ z16 are really too cluttered, no much rooms to render subway labels :cry:.

How about:
- High-speed from z11
- Main rail lines from z14
- Others from z17?

littlebtc commented:
Latest changes:
- `highspeed=yes` z11->10, with halo, z17->11, z19->12
- `usage=main` z14->10,  z17->11, z19->12
- Others z17->10, z19->11

imagico commented:
I have had a quick look at this and it seems diligent work.  I however see a lot of duplicate labels on both sides of tracks, not sure what causes this.

It would also be of interest to get the opinions of the OpenRailwayMap people - cc @rurseekatze @Nakaner.

It might be that for normal map user with no particular interest in railways labeling the physical infrastructure is not really that useful and could lead to tagging for the renderer, i.e. mappers naming tracks based on a passenger train connection using them.

Nakaner commented:
@imagico wrote:
> I have had a quick look at this and it seems diligent work. I however see a lot of duplicate labels on both sides of tracks, not sure what causes this.

That might be caused by rendering the route relations which are available as multilinestrings in the lines table?

@imagico wrote:
> It would also be of interest to get the opinions of the OpenRailwayMap people - cc @rurseekatze @Nakaner.
> 
> It might be that for normal map user with no particular interest in railways labeling the physical infrastructure is not really that useful and could lead to tagging for the renderer, i.e. mappers naming tracks based on a passenger train connection using them.

Thank you, @imagico, for notifying me. I agree with you.

I am rather sceptical regarding the rendering of `name=*` of railway tracks.

Adding more labels clutters the map. If I had a choice between labeling a street and labeling a railway track, I would prefer the street. It is more useful for orientation.

The names of railway lines are usually the names of the routes (from infrastructure point of view) because a `railway=rail/light_rail/subway` track represents two rails with sleepers and ballast and not the services using this track. The names you are going to render are in use among railway staff and railway fans (sometimes also by the locals). But often the reference numbers of the services on the track are more useful. @all Plase have a look at openrailwaymap.org, zoom in to zoom level 13/14, look around and ask yourself if you want all these names (not the numbers) be rendered on the main map.

OSM Carto dropped the rendering of `name=*` by default for everything a few year ago. That was a great decision, even for railway tracks. Before that time, people tagged something as name which they wanted to be shown on *the* map. For example, mappers added the reference numbers of all routes using the track (`name=S1, S11, S2`) or the track number (`name=track 4`) in stations as names. Both examples are abuses of the `name` tag. Neither the reference number nor the track number in conjunction with the word `track` are names. Lots of these wrong names still remain in OSM from that time).

Someone who needs a city map which shows the streets, relevant points of interest and the public transport line, should use more appropriate map.

Even if we keep the rendering of the names as proposed, the font colour, the width and the colour of the label should be reconsidered. They are difficult to read if they are located on the neighbouring track (black font on grey background). The labels of subway tracks are difficult to read, too.

`service=*` should not be labeled at all. `name=*` is used for the name of the route (infrastructure point of view) and should not be added to tracks with `service=*` at all.

To cite the Catography Guide of OpenStreetMap Carto Style:
> It's an important feedback mechanism for mappers to validate their edits and helps to prevent unfavorable fragmentation of tag use.

OpenRailwayMap serves this job for railway infrastructure tagging and uses the whole colour spectrum for that purpose. ;-) Squeezing this into a multi purpose map is difficult.

> It's a major part of the public face of OpenStreetMap, for many people the map on osm.org rendered with this style is OpenStreetMap.

Is OpenStreetMap a project of some railway nerds? No, it isn't and I am happy that it is as it is.

If you want to do something good to this map style, remove the rendering of `name` on train platforms by rendering `ref`. This is much more useful and would encourage a good tagging style.

littlebtc commented:
> The names you are going to render are in use among railway staff and railway fans (sometimes also by the locals). But often the reference numbers of the services on the track are more useful. 

Unfortunately I think this is very country-specific.
Line refs seems to be common and useful in Europe, but not so in Asia.

I live in Taiwan; Taipei MRT just introduced their latest route refs a year ago, which is based on the color.

In China, most subway systems has numbering routes, so rendering `ref` may be better -- but not the long-range railway system.

In Hong Kong, MTR has only route names and colors but no refs.

In Tokyo, JR East, major railway company in Tokyo, just introduced their ref system in 2010s, so rendering 山手線 ("Yamanote Line") is way better than its ref "JY". Even for the Subway system which has refs for decade, rendering "Marunouchi Line" is better than just "M".

**And a travel to Tokyo without railway names is painful, as railways and subways are your main way of getting around.**

> service=* should not be labeled at all. name=* is used for the name of the route (infrastructure point of view) and should not be added to tracks with service=* at all.

Agreed. I will drop `name` rendering on  `[railway = 'INT-preserved-ssy'],  [railway = 'INT-spur-siding-yard'],  [railway = 'tram-service']` on my next commit.

littlebtc commented:
> It might be that for normal map user with no particular interest in railways labeling the physical infrastructure is not really that useful and could lead to tagging for the renderer, i.e. mappers naming tracks based on a passenger train connection using them.

...I also agreed with that. In fact, in Taiwan, I had seen relation collected stations and routes based on a express train...

littlebtc commented:
https://gist.github.com/littlebtc/24bef597053015ebc160d8c1fd2f1252
Gist for images in Tokyo and Berlin, from z11 to z19.

- On z11~13, only `highspeed=yes`, so impact is extremely low.
- `usage=main` joins at z14, making major railways available.
- Others at z17 to provide more details on urban area.

kocio-pl commented:
I have nothing against hispeed names - they are not so common and may be useful hint outside urban areas. usage=main might be also OK. But looking at ORM it looks like instead of names there are sometimes refs and it's not what I would like to see on a generic map.

I would rather not render all the others.

littlebtc commented:
**Rendering `ref` instead of `name`?** 

For some region it will result in better railway rendering, and it will encourage people to add proper railway refs in some country like Japan.

If you look at OpenRailwayMap in Tokyo in z12, you will find most of the ref-ed railways is still using names:
<img width="872" alt="2017-08-04 3 16 06" src="https://user-images.githubusercontent.com/82419/28939273-54c724fe-78c3-11e7-940e-5322fd7b827e.png">

But it may hurt urban areas without any railway referencing system. :cry: So it could be a both tagging and rendering problems.

**Not rendering others at z17+**

A good question.

To me the main problem is the subway name:
1. As the current rendering focus on highway much more than railway, seeing subway name in the secondary/tertiary road feels bad
2. It will also hurt name rendering on communities with subway underneath.

I still hope it could be there, but If z17+ rendering is not eventually implementeed, I wish there are some more renderer to do that, like Transport Map.

-----------------------
I got my Docker optimization pull-request ready, so I will try export 100% images before I went to sleep.
(It is 03:27AM in Taichung)

kocio-pl commented:
I think on the general map the most important thing are the stops, stations and terminals, not the names of the lines, so I'm cautious here.

littlebtc commented:
IMHO I think the problem is that OSM-carto rendering is global, and railway statuses differ too much in countries.

For Google Maps, railways names and priorities are high only in Japan. (Now the even use totally different local map provided by Korea company in Korea)

littlebtc commented:
I pick London again as its railway network seems to be more complex than Berlin.
I pick z15~z19 as they seem to be the most controversial part.

All is using Kosmtik export with x1 scale.

z15
![london-before-z15](https://user-images.githubusercontent.com/82419/28941648-32e36ac0-78cb-11e7-9b09-7d8a3fed990c.png)
![london-after-z15](https://user-images.githubusercontent.com/82419/28941647-32d710fe-78cb-11e7-8571-9dc1204e5115.png)

z16
![london-before-z16](https://user-images.githubusercontent.com/82419/28941358-57459330-78ca-11e7-9296-c9deb43eaca0.png)
![london-after-z16](https://user-images.githubusercontent.com/82419/28941364-5d12aa96-78ca-11e7-9f22-fa1fc7c1d7cf.png)

z17
![london-before-z17](https://user-images.githubusercontent.com/82419/28941398-7efe5344-78ca-11e7-8b82-de36dfcad0f9.png)
![london-after-z17](https://user-images.githubusercontent.com/82419/28941583-13176ade-78cb-11e7-93a7-34cd81a8d08d.png)

z18
![london-before-z18](https://user-images.githubusercontent.com/82419/28941689-54173320-78cb-11e7-8d18-7452f52e1970.png)
![london-after-z18](https://user-images.githubusercontent.com/82419/28941357-573a72d4-78ca-11e7-8814-e3980b799e14.png)

z19
![london-before-z19](https://user-images.githubusercontent.com/82419/28941698-5f82ec4a-78cb-11e7-9072-564d756429bc.png)
![london-after-z19](https://user-images.githubusercontent.com/82419/28941697-5f210ea8-78cb-11e7-9214-04590edbec19.png)

littlebtc commented:
Or, can we still render other names in z17 but with very large `text-spacing` e.g. 1000?

I think it may be a good workaround, but I have to sleep now.

kocio-pl commented:
Thanks for proper rendering exports- it makes it much easier for me to compare changes!

What I think now is:
1. I want no subway line names - they look too messy even on z19.
2. I don't want any "track" names on any zoom level - platform names/refs are what we want on general map and we have them already.

Using bigger text spacing may be nice too, since it would not clutter urban areas too much, while still being visible as a hint on the outdoor.

littlebtc commented:
Trying `text-spacing=1200` and it looks much better for me in London.

For z17+, tomorrow I will try:
* highspeed and usage=main for `text-spacing`=600
* Others, `text-spacing`=1200

But I have to sleep. 🛌

littlebtc commented:
For `tracks` in the map, I think that could be a tagging question since in that station both `name` and `track_ref` are tagged but I did not seen that kind of over-tagging in other stations.

😪

Nakaner commented:
`railway:track_ref` is not the suitable tag to be rendered by a map whose users are not only railway (infrastructure) enthusiasts. There are lots of stations where the track numbers signed on the platform and the track numbers used by the infrastructure operator are different. The track numbers used by the operator of the infrastructure can be derived from the `ref=*` tags of the signals.

Rendering `railway:track_ref` will lead to situations where people will make fun about OSM because it renders "wrong numbers" and others trying to "correct" them. The only usuable reference numbers are tagged on platforms and stop positions.

Rendering the reference numbers of tracks in stations is difficult to get it right. Stations are locations where there are usually more parallel tracks than outside the stations, i.e. the labels of the tracks would be on rendered on another track and difficult to read.

littlebtc commented:
So I believe the track_ref tagging on `name` is a tagging mistake made by some user.

DerDakon commented:
If you refer to the "names" shown in the Warsaw and St. Pancras images, that just say "track N": yes, they shouldn't be there. The [OpenRailwayMap rules for JOSM validator](https://github.com/OpenRailwayMap/OpenRailwayMap/tree/master/validator) warn about this for a long time (for German, English, and French).

littlebtc commented:
First, I think current tagging of `ref` is too much different from countries (very good on Europe but not others), so I will only try to render `name` in this pull request.

A lot of changes implemented to make a better map and lesser impact:

- Bigger text-spacing:
  - `highspeed=yes`: z11=300, z14=600
  - `usage=main`: z14=300, z17=600
  - All others: z17=900
- Adding `text-min-distance: 18` for all labels to make then less crowded.
- All labels colored `#333333` -> `#666666` and default halos added.

By following is the before/after renderingfor the current result.

Now I taking London and Tokyo for example, fell free to provide a better sample area. 😄

**London**
King's Cross, z15
![london-z15-before](https://user-images.githubusercontent.com/82419/28977898-7a968748-7976-11e7-8aaf-8c7193ec54cb.png)
![london-z15-after](https://user-images.githubusercontent.com/82419/28977897-7a920b82-7976-11e7-913f-cccde051cb83.png)

King's Cross, z17
![london-z17-before](https://user-images.githubusercontent.com/82419/28977923-97424382-7976-11e7-984c-4dac7c11b1d4.png)
![london-z17-after](https://user-images.githubusercontent.com/82419/28977922-97417bdc-7976-11e7-8402-56aaf2c323d3.png)

**Tokyo**
Central Tokyo, z14
![tokyo-z14-before](https://user-images.githubusercontent.com/82419/28977966-bf9ee3da-7976-11e7-8fec-cc6757c4ef3f.png)
![tokyo-z14-after](https://user-images.githubusercontent.com/82419/28977965-bf91e7f2-7976-11e7-85d8-73e5d5825027.png)

Ueno and Akihabara, z15
![tokyo-z15-before](https://user-images.githubusercontent.com/82419/28978211-ad2ed646-7977-11e7-9102-1be5301a01b9.png)
![tokyo-z15-after](https://user-images.githubusercontent.com/82419/28978210-ad2e9cda-7977-11e7-91bb-200c597f8fc4.png)

Ueno, z17
![tokyo-z17-before](https://user-images.githubusercontent.com/82419/28978308-13dbb0da-7978-11e7-9835-2af28515e243.png)
![tokyo-z17-after](https://user-images.githubusercontent.com/82419/28978307-1398d792-7978-11e7-91c9-0f924c99e20e.png)


-------------
To me, central Tokyo improvement on z14 is very impressive 😃

kocio-pl commented on roads.mss:
```diff
@@ -3106,3 +3106,111 @@ tertiary is rendered from z10 and is not included in osm_planet_roads. */
+    }
+  }
+  /* Other minor railway styles.
+     For service rails, see #2687. */
```
I would give full URL here:
https://github.com/gravitystorm/openstreetmap-carto/pull/2687

kocio-pl commented:
Much better version, indeed. I have spotted only single cases where railway labels are covering some other labels, which is very good in such big cities.

Hi-speed in Poland on z11 looks reasonable (spacing could be bigger, but changing this option doesn't work here because of many segments): 
![t5cyotvm](https://user-images.githubusercontent.com/5439713/28999838-cda41ea0-7a57-11e7-835f-3cfe4e8f5949.png)

Thanks to the color change railway=rail labels on z17 are no longer competing with street names:
![meobvaks](https://user-images.githubusercontent.com/5439713/28999932-2bd2a94e-7a5b-11e7-9e1c-f6b08449feca.png)

When the named road and tram line are shared, the rail label is sometimes visible. Road name takes precedence, which makes the problem less severe, but we should not render railway label at all in such cases:
![qpnv2cnk](https://user-images.githubusercontent.com/5439713/29000046-9cbb2560-7a5f-11e7-9466-3c3cc2a03604.png)
![19rmgvr_](https://user-images.githubusercontent.com/5439713/29000107-ebe87a50-7a61-11e7-8a36-e86ad8248469.png)

DerDakon commented:
If the same way is used for both the highway and the tram then the railway label simply should never be shown. It's probably not the correct name for the railway anyway.

kocio-pl commented:
Try to use z_order instead of priority table, see https://github.com/gravitystorm/openstreetmap-carto/pull/2732.

littlebtc commented:
Review comments committed.

----------
To me, the major concern is the railway names rendered bad in multiple-track railway mixed together:
![bad_railway_name_rendering](https://user-images.githubusercontent.com/82419/29083697-266fccb4-7c9c-11e7-91f6-893881f1eac7.png)

Is there any way to improve rendering like this case?

----------
Another confusing issue is that in the current rendering, subway is almost invisible under highway, but names from subway would show. Can we improve subway rendering a little bit?

kocio-pl commented on roads.mss:
```diff
@@ -3106,3 +3106,111 @@ tertiary is rendered from z10 and is not included in osm_planet_roads. */
+  [railway = 'monorail'],
+  [railway = 'turntable'],
+  [railway = 'tram'] {
+    [zoom >= 17] {
```
You need to add something like [highway = null] to avoid road names repeated as railway labels.

littlebtc replied:
Running this on SQL side may be better?

kocio-pl replied:
I don't know, but probably yes.

littlebtc commented:
I try to change all `text-dy` to 4 as railways are much lighter thant highway.

Testing how this will effect rendering.

kocio-pl commented:
You can drop here just some interesting rendering samples, full overview would be too much images and not necessary.

littlebtc commented:
Reverted that approach as it seems too crowded.

littlebtc commented:
Export feature from kosmtik seems to be broken...

After several approach and failure, I had only changed the `text-dy` for `highspeed=yes`. z11 and z12 now has `text-dy=3` as the rails are relativlely think in these two zoom level.

![2017-08-11 11-42-11](https://user-images.githubusercontent.com/82419/29200982-9d2c8b20-7e8c-11e7-9c2e-6e4f350ab813.png)
![2017-08-11 11-42-26](https://user-images.githubusercontent.com/82419/29200984-9d341bc4-7e8c-11e7-8796-2d2531f64467.png)
![2017-08-11 11-46-52](https://user-images.githubusercontent.com/82419/29200983-9d2f600c-7e8c-11e7-9395-d30ee0c77826.png)

Other rendering problems seem too hard to have a easy fix.

DerDakon commented on roads.mss:
```diff
@@ -3106,3 +3106,114 @@ tertiary is rendered from z10 and is not included in osm_planet_roads. */
+  [railway = 'preserved'],
+  [railway = 'funicular'],
+  [railway = 'monorail'],
+  [railway = 'turntable'],
```
A turntable is something like the railway equivalent of a parking lane crossing or so. I don't think any of them can have a useful name.

kocio-pl replied:
Sure, it doesn't belong here.

eruedin commented:
This looks great, but not for a general map. If it is for the public transport map or for the railway map, that  are the information I am looking for, but not on the general map. I think, that is the reason, why we have a general map, a bicycle map and a public transport map. They show different things for different needs.

kocio-pl commented:
I'm inclined to have this code. It was too visible in the first draft, but has been tuned a lot in subsequent steps and it's not disturbing for me any more. 

Of course it would be out of question to show everything like OpenRailwayMap or as prominently as the transport layer. Refs don't belong here for example. But names are what people can identify - for example some rail lines are touristic attractions or locally important transport brand (including trams) and hi-speed lines are special feature worth showing in the outdoor. I would however not show any service lines.

I still hesitate when it comes to underground lines. I see some analogy to quarry - we show only overground buildings and areas. For underground passengers stations (and even entry numbers) is enough I guess.

kocio-pl commented:
I think that names in tunnels also shouldn't be rendered, because we can't avoid situations like this:
![9arevcdw](https://user-images.githubusercontent.com/5439713/29438648-59bf054e-83b8-11e7-836b-03f7f124c6e3.png)

math1985 commented:
I tested this, it works as described. I'm fine with merging it (once @kocio-pl's remarks are taken into account).

littlebtc commented:
Comments addressed in latest pushes.

kocio-pl commented:
Subway will cause the same problems as tunnels, so should be removed too.

And the last thing - please join all the commits into just one.

littlebtc commented:
Some subway is ground-leveled or elevated, so I think subway is OK.

2017年8月19日 下午6:57，"kocio-pl" <notifications@github.com>寫道：

> Subway will cause the same problems as tunnels, so should be removed too.
>
> And the last thing - please join all the commits into just one.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gravitystorm/openstreetmap-carto/pull/2687#issuecomment-323516259>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFB89bcfS7S8TBTOHc0o2cXEgZ1j8azks5sZr-CgaJpZM4OUx-a>
> .
>

kocio-pl commented:
I'm aware if this, but I guess they are under ground most of the time. It's not about just trains - I think about [hiding most level<>0 objects](https://github.com/gravitystorm/openstreetmap-carto/issues/1977).

littlebtc commented:
Rebased and squashed all commits into one.

For subway I think current status should be better, since there should have lots of >5km elevated/ground-level subways.

pnorman commented:
Subway should have tunnel if it's below ground. I'd wouldn't rule out showing underground subway line names, but am not sure. Probably best to treat them the same for the initial version.

kocio-pl commented:
OK, you're right. I guess there will be some work to be done, like moving false names into proper refs and maybe some other issues to be fixed later, but we'll see.

kocio-pl commented:
Thanks a lot for this PR and the work with tuning it!
